private/bufpkg/bufmodule/errors.go

```diff
@@ -209,3 +212,47 @@ func (n *NoProtoFilesError) Error() string {
 	_, _ = builder.WriteString(` had no .proto files`)
 	return builder.String()
 }
+
+// DigestMismatchError is the error returned if the Digest of a downloaded Module or Commit
+// does not match the expected digest in a buf.lock file.
+type DigestMismatchError struct {
+	ModuleFullName ModuleFullName
+	CommitID       uuid.UUID
+	ExpectedDigest Digest
+	ActualDigest   Digest
+}
+
+// Error implements the error interface.
+func (m *DigestMismatchError) Error() string {
+	if m == nil {
+		return ""
+	}
+	var builder strings.Builder
+	_, _ = builder.WriteString(`*** Digest verification failed`)
+	if m.ModuleFullName != nil {
+		_, _ = builder.WriteString(` for "`)
+		_, _ = builder.WriteString(m.ModuleFullName.String())
+		if !m.CommitID.IsNil() {
+			_, _ = builder.WriteString(`:`)
+			_, _ = builder.WriteString(uuidutil.ToDashless(m.CommitID))
+		}
+		_, _ = builder.WriteString(`"`)
+	}
+	_, _ = builder.WriteString(` ***`)
+	_, _ = builder.WriteString("\n")
+	if m.ExpectedDigest != nil && m.ActualDigest != nil {
+		_, _ = builder.WriteString("\t")
+		_, _ = builder.WriteString(`Expected digest: "`)
+		_, _ = builder.WriteString(m.ExpectedDigest.String())
+		_, _ = builder.WriteString(`"`)
+		_, _ = builder.WriteString("\n")
+		_, _ = builder.WriteString("\t")
+		_, _ = builder.WriteString(`Actual digest: "`)
```

Comment:
Do we have information to differentiate if the actual digest is from the module cache or the BSR here?

If we fetch the digest from the BSR we could perhaps provide a better error message here to the user to choose a better message ('clear cache' vs. 'attack').

Reply:
We don't at the moment, we can do this later if we need to